mobile_config/src/gateway/db.rs

```diff
@@ -0,0 +1,365 @@
+                elevation,
+                azimuth,
+                location,
+                location_changed_at,
+                location_asserts
+            )
+            VALUES (
+                $1, $2, $3, $4, $5, $6, $7,
+                $8, $9, $10, $11, $12, $13
+            )
+            ON CONFLICT (address)
+            DO UPDATE SET
+                gateway_type = EXCLUDED.gateway_type,
+                created_at = EXCLUDED.created_at,
+                updated_at = EXCLUDED.updated_at,
+                refreshed_at = EXCLUDED.refreshed_at,
+                last_changed_at = CASE
+                    WHEN gateways.location IS DISTINCT FROM EXCLUDED.location
+                        OR gateways.hash IS DISTINCT FROM EXCLUDED.hash
+                    THEN EXCLUDED.refreshed_at
+                    ELSE gateways.last_changed_at
+                END,
+                hash = EXCLUDED.hash,
+                antenna = EXCLUDED.antenna,
+                elevation = EXCLUDED.elevation,
+                azimuth = EXCLUDED.azimuth,
+                location = EXCLUDED.location,
+                location_changed_at = CASE
+                    WHEN gateways.location IS DISTINCT FROM EXCLUDED.location
+                    THEN EXCLUDED.refreshed_at
+                    ELSE gateways.location_changed_at
+                END,
+                location_asserts = EXCLUDED.location_asserts
+            "#,
+        )
+        .bind(self.address.as_ref())
+        .bind(self.gateway_type)
+        .bind(self.created_at)
+        .bind(self.updated_at)
+        .bind(self.refreshed_at)
+        .bind(self.last_changed_at)
+        .bind(self.hash.as_str())
+        .bind(self.antenna.map(|v| v as i64))
+        .bind(self.elevation.map(|v| v as i64))
+        .bind(self.azimuth.map(|v| v as i64))
+        .bind(self.location.map(|v| v as i64))
+        .bind(self.location_changed_at)
+        .bind(self.location_asserts.map(|v| v as i64))
+        .execute(pool)
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn get_by_address<'e>(
+        db: impl PgExecutor<'e>,
+        address: &PublicKeyBinary,
+    ) -> anyhow::Result<Option<Self>> {
+        let gateway = sqlx::query_as::<_, Self>(
+            r#"
+            SELECT
+                address,
+                gateway_type,
+                created_at,
+                updated_at,
+                refreshed_at,
+                last_changed_at,
+                hash,
+                antenna,
+                elevation,
+                azimuth,
+                location,
+                location_changed_at,
+                location_asserts
+            FROM gateways
+            WHERE address = $1
+            "#,
+        )
+        .bind(address.as_ref())
+        .fetch_optional(db)
+        .await?;
+
+        Ok(gateway)
+    }
+
+    pub fn stream_by_addresses<'a>(
+        db: impl PgExecutor<'a> + 'a,
+        addresses: &[PublicKeyBinary],
```

Comment:
addresses could be `Vec<PublicKeyBinary>`. 

Everywhere this function is called, the addresses are never re-used. 
Ownership could be given, instead of cloning the contents of the slice just for the source to be dropped without doing anything else.

And the first line could be
```rs
let addr_array: Vec<Vec<u8>> = addresses.into_iter().map(|a| a.into()).collect();
```

Reply:
That ownership could flow up through `gateway::service::info::stream_by_addresses()` as well.